google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java

```diff
@@ -0,0 +1,3482 @@
+      return TimestampBound.ofExactStaleness(
+          (long) (concurrency.getStalenessSeconds() * 1000000), TimeUnit.MICROSECONDS);
+    } else if (concurrency.hasMinReadTimestampMicros()) {
+      return TimestampBound.ofMinReadTimestamp(
+          timestampFromMicros(concurrency.getMinReadTimestampMicros()));
+    } else if (concurrency.hasMaxStalenessSeconds()) {
+      return TimestampBound.ofMaxStaleness(
+          (long) (concurrency.getMaxStalenessSeconds() * 1000000), TimeUnit.MICROSECONDS);
+    } else if (concurrency.hasExactTimestampMicros()) {
+      return TimestampBound.ofReadTimestamp(
+          timestampFromMicros(concurrency.getExactTimestampMicros()));
+    } else if (concurrency.hasStrong()) {
+      return TimestampBound.strong();
+    } else if (concurrency.hasBatch()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT,
+          "batch mode should not be in snapshot transaction: " + concurrency);
+    }
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.INVALID_ARGUMENT, "Unsupported concurrency mode: " + concurrency);
+  }
+
+  /** Get DatabaseId by parsing the given dbPath. */
+  private static DatabaseId getDatabaseId(String dbPath) {
+    Matcher matcher = DB_NAME.matcher(dbPath);
+    if (!matcher.matches()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Bad database path: " + dbPath);
+    }
+    return DatabaseId.of(matcher.group(1), matcher.group(2), matcher.group(3));
+  }
+
+  /** Build instance proto from cloud spanner instance. */
+  private com.google.spanner.admin.instance.v1.Instance instanceToProto(Instance instance) {
+    com.google.spanner.admin.instance.v1.Instance.Builder instanceBuilder =
+        com.google.spanner.admin.instance.v1.Instance.newBuilder();
+    instanceBuilder
+        .setConfig(instance.getInstanceConfigId().getInstanceConfig())
+        .setName(instance.getId().getName())
+        .setDisplayName(instance.getDisplayName())
+        .setCreateTime(instance.getCreateTime().toProto())
+        .setNodeCount(instance.getNodeCount())
+        .setProcessingUnits(instance.getProcessingUnits())
+        .setUpdateTime(instance.getUpdateTime().toProto())
+        .putAllLabels(instance.getLabels());
+    com.google.spanner.admin.instance.v1.Instance.State state;
+    switch (instance.getState()) {
+      case UNSPECIFIED:
+        state = State.STATE_UNSPECIFIED;
+        break;
+      case CREATING:
+        state = State.CREATING;
+        break;
+      case READY:
+        state = State.READY;
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown state:" + instance.getState());
+    }
+    instanceBuilder.setState(state);
+    return instanceBuilder.build();
+  }
+
+  /** Build instance proto from cloud spanner instance. */
+  private com.google.spanner.admin.instance.v1.InstanceConfig instanceConfigToProto(
```

Comment:
Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java#L418?

Reply:
This API is not public.